File: libpapyro/papyro/cslengine/cslengine.js
/*****************************************************************************
 *  
 *   This file is part of the Utopia Documents application.
 *       Copyright (c) 2008-2017 Lost Island Labs
 *           <info@utopiadocs.com>
 *   
 *   Utopia Documents is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU GENERAL PUBLIC LICENSE VERSION 3 as
 *   published by the Free Software Foundation.
 *   
 *   Utopia Documents is distributed in the hope that it will be useful, but
 *   WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 *   Public License for more details.
 *   
 *   In addition, as a special exception, the copyright holders give
 *   permission to link the code of portions of this program with the OpenSSL
 *   library under certain conditions as described in each individual source
 *   file, and distribute linked combinations including the two.
 *   
 *   You must obey the GNU General Public License in all respects for all of
 *   the code used other than OpenSSL. If you modify file(s) with this
 *   exception, you may extend this exception to your version of the file(s),
 *   but you are not obligated to do so. If you do not wish to do so, delete
 *   this exception statement from your version.
 *   
 *   You should have received a copy of the GNU General Public License
 *   along with Utopia Documents. If not, see <http://www.gnu.org/licenses/>
 *  
 *****************************************************************************/

//
// This CSL engine wrapper works with the CSLEngine class to set up and provide CSL
// processing funationality from within Qt.
//




// Hack to fix a couple of bugs in citeproc
// Alias the XML language attribute
CSL.Attributes["@xml:lang"] = function (state, arg) {
    // Add a dummy version in if it isn't already present
    if (!state.opt.version) {
        state.opt.version = '1.0';
    }
    // Call the old callback
    var csl_lang_fn = CSL.Attributes["@lang"];
    csl_lang_fn(state, arg);
}




// Wrapper state for keeping track of styles / locales
var Utopia = {
    styles: {},
    locales: {},
    defaultStyle: undefined,
};

function installStyle(code, name, style)
{
    // Find the title from the style JSON
    function find(obj, name) {
        if (obj !== null && typeof obj === 'object' && obj.children) {
            var path = name.split('/');
            name = path.shift();
            if (name == 'text()') {
                return obj.children.join(' ');
            } else {
                for (i = 0; i < obj.children.length; i += 1) {
                    if (obj.children[i].name == name) {
                        if (path.length > 0) {
                            return find(obj.children[i], path.join('/'));
                        } else {
                            return obj.children[i];
                        }
                    }
                }
            }
        }
    }

    Utopia.styles[code] = {
        code: code,
        name: find(style, 'info/title/text()'),
        json: style,
    };
}

function installLocale(code, name, locale)
{
    Utopia.locales[code] = {
        code: code,
        name: name,
        json: locale,
    };
}

function getStyles()
{
    var styles = {};
    for (var code in Utopia.styles) {
        styles[code] = Utopia.styles[code].name;
    }
    return styles;
}

function getLocales()
{
    var locales = {};
    for (var code in Utopia.locales) {
        locales[code] = Utopia.locales[code].name;
    }
    return locales;
}




function convert_name(name)
{
    // From ["Thorne, David"] to [{family: "Thorne", given: "David"}]
}

function convert_date(year, month, day)
{
    // From [2014, March, 12] to {date-parts: [[2014, 3, 12]]}
}

function get_case_insensitive(map, needle) {
    var keys = Object.keys(map);
    for (i = 0; i < keys.length; i += 1) {
        var key = keys[i];
        if (key.toLowerCase() == needle.toLowerCase()) {
            return map[key];
        }
    }
    return false;
}

function format(metadata, style, defaultStyle)
{
    // Give the metadata an ID if it doesn't already have one
    if (!metadata.id) {
        if (metadata.label) {
            metadata.id = metadata.label;
        } else {
            metadata.id = '';
        }
    }

    // Resolve style
    found = get_case_insensitive(Utopia.styles, style) ||
            get_case_insensitive(Utopia.styles, defaultStyle) ||
            get_case_insensitive(Utopia.styles, 'apa');
    if (found) {
        style = found.json;
    }

    var sys = {
        retrieveLocale: function (name) {
            if (Utopia.locales[name]) {
                return Utopia.locales[name].json;
            }
        },
        retrieveItem: function (id) {
            return metadata;
        },
        getAbbreviations: function (name) {
            return {};
        },
    };

    var label = metadata['citation-label'];

    var citeproc = new CSL.Engine(sys, style, 'en-GB')
    citeproc.updateItems([metadata.id], true);
    var bib = citeproc.makeBibliography("CITATION_LABEL");
    var formatted = bib[1][0];

    // Put label in place
    if ("string" === typeof label) {
        if (formatted.indexOf("CITATION_LABEL") == -1) {
            formatted = 'CITATION_LABEL ' + formatted;
        }
        formatted = formatted.replace('CITATION_LABEL', '<strong>' + label.replace(/\.+$/g, '') + '</strong>');
    } else {
        formatted = formatted.replace(/[^a-zA-Z>]*CITATION_LABEL[^a-zA-Z<]*/, '');
    }

    // Swap divs for spans
    formatted = formatted.replace(/<div/g, '<span').replace(/<\/div/g, '</span');

    return formatted;
}

File: libpapyro/papyro/cslengine.cpp
/*****************************************************************************
 *  
 *   This file is part of the Utopia Documents application.
 *       Copyright (c) 2008-2017 Lost Island Labs
 *           <info@utopiadocs.com>
 *   
 *   Utopia Documents is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU GENERAL PUBLIC LICENSE VERSION 3 as
 *   published by the Free Software Foundation.
 *   
 *   Utopia Documents is distributed in the hope that it will be useful, but
 *   WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 *   Public License for more details.
 *   
 *   In addition, as a special exception, the copyright holders give
 *   permission to link the code of portions of this program with the OpenSSL
 *   library under certain conditions as described in each individual source
 *   file, and distribute linked combinations including the two.
 *   
 *   You must obey the GNU General Public License in all respects for all of
 *   the code used other than OpenSSL. If you modify file(s) with this
 *   exception, you may extend this exception to your version of the file(s),
 *   but you are not obligated to do so. If you do not wish to do so, delete
 *   this exception statement from your version.
 *   
 *   You should have received a copy of the GNU General Public License
 *   along with Utopia Documents. If not, see <http://www.gnu.org/licenses/>
 *  
 *****************************************************************************/

#include <papyro/cslengine.h>
#include <utopia2/global.h>

#include <boost/weak_ptr.hpp>

#include <QDir>
#include <QDomDocument>
#include <QFile>
#include <QJsonDocument>
#include <QMutex>
#include <QMutexLocker>
#include <QResource>
#include <QScriptEngine>
#include <QSettings>
#include <QString>
#include <QStringList>
#include <QVariantMap>

#include <QDebug>

namespace Papyro
{

    namespace
    {

        QVariant elementToVariant(const QDomElement & element)
        {
            QVariantMap obj;
            obj["name"] = element.tagName();
            QVariantMap attrs;
            QDomNamedNodeMap attributes(element.attributes());
            for (int i = 0; i < attributes.count(); ++i) {
                QDomAttr attrNode(attributes.item(i).toAttr());
                attrs[attrNode.name()] = attrNode.value();
            }
            obj["attrs"] = attrs;
            QVariantList children;
            if (element.childNodes().size() == 0 && element.tagName() == "term") {
                children << QString("");
            }
            for(QDomNode n = element.firstChild(); !n.isNull(); n = n.nextSibling()) {
                if (n.isText()) {
                    children << n.toText().data();
                } else if (n.isElement()) {
                    children << elementToVariant(n.toElement());
                }
            }
            obj["children"] = children;
            return obj;
        }

        QString xmlToJson(const QString & xml)
        {
            QVariantMap map;
            QDomDocument doc;
            doc.setContent(xml.toUtf8());

            QDomElement elem = doc.documentElement();
            QVariant converted = elementToVariant(elem);

            return QString::fromUtf8(QJsonDocument::fromVariant(converted).toJson());
        }

        QScriptValue retrieveLocale(QScriptContext * context, QScriptEngine * engine)
        {
            if (context->argumentCount() != 1) {
                return context->throwError("retrieveLocale() takes exactly one argument");
            }

            QString lang = context->argument(0).toString();
            QString xml;

            QFile citeproc(Utopia::resource_path() + "/citeproc/locales/" + lang + ".json");
            if (citeproc.open(QIODevice::ReadOnly)) {
                QByteArray raw(citeproc.readAll());
                xml = QString::fromUtf8(raw.constData(), raw.size());
            }

            return engine->evaluate("(" + xml + ")", Utopia::resource_path() + "/citeproc/locales/" + lang + ".json");
        }

        QScriptValue debug(QScriptContext * context, QScriptEngine * engine)
        {
            if (context->argumentCount() != 1) {
                return context->throwError("debug() takes exactly one argument");
            }

            qDebug() << context->argument(0).toString();

            return engine->undefinedValue();
        }

        QString resource(const QString & path, bool warn = true)
        {
            QFile file(path);
            if (file.open(QIODevice::ReadOnly)) {
                QByteArray raw(file.readAll());
                return QString::fromUtf8(raw.constData(), raw.size());
            } else if (warn) {
                qDebug() << "ERROR in ::resource(): Cannot load " + path;
            }
            return QString();
        }

        QScriptValue evaluate(QScriptEngine * engine, const QString & script, const QString & fileName = QString())
        {
            QScriptValue ret = engine->evaluate(script, fileName);
            if (engine->hasUncaughtException()) {
                qDebug() << "EXCEPTION";
                qDebug() << engine->uncaughtException().toString();
                foreach (QString line, engine->uncaughtExceptionBacktrace()) {
                    qDebug() << line;
                }
                engine->clearExceptions();
            }
            return ret;
        }

    }




    class CSLEnginePrivate
    {
    public:
        CSLEnginePrivate()
            : mutex(QMutex::Recursive)
        {
            // Populate from settings
            QSettings conf;
            conf.sync();
            conf.beginGroup("CSLEngine");
            defaultStyle = conf.value("Default Style", "apa").toString();

            // Provide the environment with helper methods
            QScriptValue globalObject = engine.globalObject();
            globalObject.setProperty(QString("retrieveLocale"), engine.newFunction(retrieveLocale));
            globalObject.setProperty(QString("debug"), engine.newFunction(debug));

            // Install javascript control code
            QStringList sources;
            sources << ":/citeproc/xmljson.js";
            sources << ":/citeproc/citeproc.js";
            sources << ":/cslengine.js";
            foreach (const QString & source, sources) {
                evaluate(&engine, resource(source), source);
            }

            // Grab handle for admin functions
            QScriptValue installLocaleFn = globalObject.property("installLocale");
            QScriptValue installStyleFn = globalObject.property("installStyle");

            // Install locales
            QVariantMap localeMap(evaluate(&engine, "(" + resource(Utopia::resource_path() + "/citeproc/locales.json") + ")", Utopia::resource_path() + "/citeproc/locales.json").toVariant().toMap());
            QMapIterator< QString, QVariant > localeMapIter(localeMap);
            while (localeMapIter.hasNext()) {
                localeMapIter.next();
                QString code = localeMapIter.key();
                QString description = localeMapIter.value().toString();
                if (code != "description") {
                    QString locale = resource(Utopia::resource_path() + "/citeproc/locales/" + code + ".json", false);
                    if (!locale.isEmpty()) {
                        QScriptValueList args;
                        args << engine.toScriptValue(code);
                        args << engine.toScriptValue(description);
                        args << engine.evaluate("(" + locale + ")");
                        installLocaleFn.call(globalObject, args);
                        qDebug() << "CSLEngine: Loaded locale" << code;
                    }
                }
            }

            // Install styles
            QDir stylesDir(Utopia::resource_path() + "/citeproc/styles");
            QDir userStylesDir(Utopia::profile_path() + "/csl");
            QStringList filters;
            filters << "*.json" << "*.csl";
            stylesDir.setNameFilters(filters);
            userStylesDir.setNameFilters(filters);
            QFileInfoList styleFiles = stylesDir.entryInfoList() + userStylesDir.entryInfoList();
            foreach (const QFileInfo & styleFileInfo, styleFiles) {
                QString styleFile = styleFileInfo.fileName();
                QString code = styleFile.section(".", 0, 0);
                QString description(code);
                description = description.replace(QRegExp("[^a-zA-Z0-9]+"), " ").trimmed();
                QString style = resource(styleFileInfo.filePath());
                if (styleFileInfo.suffix() == "csl") {
                    style = xmlToJson(style);
                }
                if (!style.isEmpty()) {
                    QScriptValueList args;
                    args << engine.toScriptValue(code);
                    args << engine.toScriptValue(description);
                    args << engine.evaluate("(" + style + ")");
                    installStyleFn.call(globalObject, args);
                    qDebug() << "CSLEngine: Loaded style" << styleFile;
                } else {
                    qDebug() << "CSLEngine: Could not load style" << styleFile;
                }
            }
        }

        ~CSLEnginePrivate()
        {}

        QString defaultStyle;
        QScriptEngine engine;
        QMutex mutex;

    }; // class CSLEnginePrivate




    CSLEngine::CSLEngine(QObject * parent)
        : QObject(parent), d(new CSLEnginePrivate)
    {}

    CSLEngine::~CSLEngine()
    {
        delete d;
    }

    QVariantMap CSLEngine::availableLocales() const
    {
        QMutexLocker guard(&d->mutex);
        QVariantMap locales;
        QScriptValue globalObject = d->engine.globalObject();
        QScriptValue getLocalesFn = globalObject.property("getLocales");
        if (getLocalesFn.isFunction()) {
            QVariantMap map(getLocalesFn.call(globalObject).toVariant().toMap());
            if (!d->engine.hasUncaughtException()) {
                QMapIterator< QString, QVariant > iter(map);
                while (iter.hasNext()) {
                    iter.next();
                    locales[iter.key()] = iter.value();
                }
            } else {
                qDebug() << "EXCEPTION";
                qDebug() << d->engine.uncaughtException().toString();
                foreach (QString line, d->engine.uncaughtExceptionBacktrace()) {
                    qDebug() << line;
                }
                d->engine.clearExceptions();
            }
        }
        return locales;
    }

    QVariantMap CSLEngine::availableStyles() const
    {
        QMutexLocker guard(&d->mutex);
        QVariantMap styles;
        QScriptValue globalObject = d->engine.globalObject();
        QScriptValue getStylesFn = globalObject.property("getStyles");
        if (getStylesFn.isFunction()) {
            QVariantMap map(getStylesFn.call(globalObject).toVariant().toMap());
            if (!d->engine.hasUncaughtException()) {
                QMapIterator< QString, QVariant > iter(map);
                while (iter.hasNext()) {
                    iter.next();
                    styles[iter.key()] = iter.value();
                }
            } else {
                qDebug() << "EXCEPTION";
                qDebug() << d->engine.uncaughtException().toString();
                foreach (QString line, d->engine.uncaughtExceptionBacktrace()) {
                    qDebug() << line;
                }
                d->engine.clearExceptions();
            }
        }
        return styles;
    }

    const QString & CSLEngine::defaultStyle() const
    {
        QMutexLocker guard(&d->mutex);
        return d->defaultStyle;
    }

    QString CSLEngine::format(const QVariantMap & metadata, const QString & style)
    {
        QMutexLocker guard(&d->mutex);
        QString formatted;
        QScriptValue globalObject = d->engine.globalObject();
        QScriptValue formatFn = globalObject.property("format");
        if (formatFn.isFunction()) {
            QScriptValueList args;
            args << d->engine.toScriptValue(metadata);
            args << d->engine.toScriptValue(style);
            args << d->engine.toScriptValue(defaultStyle());

            QScriptValue val = formatFn.call(globalObject, args);
            if (!d->engine.hasUncaughtException()) {
                formatted = val.toString();
            } else {
                qDebug() << "EXCEPTION ---" << val.toString();;
                qDebug() << d->engine.uncaughtException().toString();
                foreach (QString line, d->engine.uncaughtExceptionBacktrace()) {
                    qDebug() << line;
                }
                d->engine.clearExceptions();
            }
        } else {
            qDebug() << "ERROR: format doesn't seem to be a function object";
        }
        return formatted.trimmed();
    }

    boost::shared_ptr< CSLEngine > CSLEngine::instance()
    {
        static boost::weak_ptr< CSLEngine > singleton;
        boost::shared_ptr< CSLEngine > shared(singleton.lock());
        if (singleton.expired()) {
            shared = boost::shared_ptr< CSLEngine >(new CSLEngine());
            singleton = shared;
        }
        return shared;
    }

    void CSLEngine::setDefaultStyle(const QString & defaultStyle)
    {
        QMutexLocker guard(&d->mutex);
        d->defaultStyle = defaultStyle;
    }

} // namespace Papyro
